File: Phaser/Zelda/code/player.py
"""
Player class - handles Link's movement, combat, and interactions.
"""

import pygame
from settings import *


class Player(pygame.sprite.Sprite):
    def __init__(self, pos, groups, obstacle_sprites, player_data,
                 create_attack, create_projectile):
        super().__init__(groups)

        # References
        self.obstacle_sprites = obstacle_sprites
        self.player_data = player_data
        self.create_attack = create_attack
        self.create_projectile = create_projectile

        # Graphics
        self.image = pygame.Surface((TILE_SIZE * SCALE, TILE_SIZE * SCALE))
        self.image.fill(GREEN)
        self.rect = self.image.get_rect(topleft=pos)
        self.hitbox = self.rect.inflate(-8, -8)

        # Movement
        self.direction = pygame.math.Vector2()
        self.facing = player_data.get('direction', DOWN)
        self.speed = PLAYER_SPEED

        # Animation
        self.animations = self.load_animations()
        self.animation_index = 0
        self.animation_speed = ANIMATION_SPEED
        self.last_animation_time = 0
        self.status = 'idle'

        # Stats
        self.health = player_data['health']
        self.max_health = player_data['max_health']
        self.magic = player_data['magic']
        self.max_magic = player_data['max_magic']

        # Combat
        self.attacking = False
        self.attack_cooldown = SWORD_COOLDOWN
        self.attack_time = 0
        self.spin_attack_charging = False
        self.spin_charge_start = 0

        # Invincibility frames
        self.invincible = False
        self.invincible_time = 0
        self.invincibility_duration = INVINCIBILITY_DURATION

        # Item usage
        self.can_use_item = True
        self.item_cooldown = 300
        self.item_use_time = 0

        # Running (Pegasus Boots)
        self.running = False
        self.run_speed = PLAYER_SPEED * 2

        # Input state
        self.keys_pressed = {
            'up': False,
            'down': False,
            'left': False,
            'right': False,
            'action': False,
        }

    @property
    def direction_facing(self):
        """Return the direction enum for current facing."""
        return self.facing

    def load_animations(self):
        """Load player animations (placeholder - creates colored rectangles)."""
        animations = {
            'idle_down': [],
            'idle_up': [],
            'idle_left': [],
            'idle_right': [],
            'walk_down': [],
            'walk_up': [],
            'walk_left': [],
            'walk_right': [],
            'attack_down': [],
            'attack_up': [],
            'attack_left': [],
            'attack_right': [],
        }

        # Create placeholder frames
        colors = {
            'down': GREEN,
            'up': (0, 200, 0),
            'left': (0, 150, 0),
            'right': (0, 100, 0),
        }

        for direction, color in colors.items():
            # Idle frames (1 frame)
            surf = pygame.Surface((TILE_SIZE * SCALE, TILE_SIZE * SCALE))
            surf.fill(color)
            # Draw direction indicator
            pygame.draw.circle(surf, WHITE,
                               (TILE_SIZE * SCALE // 2, TILE_SIZE * SCALE // 4)
                               if direction in ['up', 'down'] else
                               (TILE_SIZE * SCALE // 4 if direction == 'left' else TILE_SIZE * SCALE * 3 // 4,
                                TILE_SIZE * SCALE // 2),
                               4)
            animations[f'idle_{direction}'] = [surf]

            # Walk frames (4 frames)
            for i in range(4):
                walk_surf = pygame.Surface((TILE_SIZE * SCALE, TILE_SIZE * SCALE))
                walk_surf.fill(color)
                # Animate legs
                leg_offset = (i % 2) * 4
                pygame.draw.rect(walk_surf, (0, 50, 0),
                                 (8, TILE_SIZE * SCALE - 12 + leg_offset, 8, 8))
                pygame.draw.rect(walk_surf, (0, 50, 0),
                                 (TILE_SIZE * SCALE - 16, TILE_SIZE * SCALE - 12 - leg_offset, 8, 8))
                # Direction indicator
                pygame.draw.circle(walk_surf, WHITE,
                                   (TILE_SIZE * SCALE // 2, TILE_SIZE * SCALE // 4)
                                   if direction in ['up', 'down'] else
                                   (TILE_SIZE * SCALE // 4 if direction == 'left' else TILE_SIZE * SCALE * 3 // 4,
                                    TILE_SIZE * SCALE // 2),
                                   4)
                animations[f'walk_{direction}'].append(walk_surf)

            # Attack frames (3 frames)
            for i in range(3):
                attack_surf = pygame.Surface((TILE_SIZE * SCALE, TILE_SIZE * SCALE))
                attack_surf.fill(color)
                # Draw sword
                sword_color = (192, 192, 192)  # Silver
                if direction == 'down':
                    pygame.draw.rect(attack_surf, sword_color,
                                     (TILE_SIZE * SCALE // 2 - 2 + (i - 1) * 8,
                                      TILE_SIZE * SCALE - 8, 4, 12))
                elif direction == 'up':
                    pygame.draw.rect(attack_surf, sword_color,
                                     (TILE_SIZE * SCALE // 2 - 2 + (i - 1) * 8, -4, 4, 12))
                elif direction == 'left':
                    pygame.draw.rect(attack_surf, sword_color,
                                     (-4, TILE_SIZE * SCALE // 2 - 2 + (i - 1) * 8, 12, 4))
                elif direction == 'right':
                    pygame.draw.rect(attack_surf, sword_color,
                                     (TILE_SIZE * SCALE - 8,
                                      TILE_SIZE * SCALE // 2 - 2 + (i - 1) * 8, 12, 4))
                animations[f'attack_{direction}'].append(attack_surf)

        return animations

    def handle_key_down(self, key):
        """Handle key press events."""
        # Sword attack
        if key in [pygame.K_j, pygame.K_z]:
            self.attack()

        # Item use
        if key in [pygame.K_k, pygame.K_x]:
            self.use_item()

        # Action button (run with Pegasus Boots)
        if key in [pygame.K_l, pygame.K_c]:
            if self.player_data['items']['boots']:
                self.running = True

        # Quick item select
        item_keys = {
            pygame.K_1: 'bombs',
            pygame.K_2: 'bow',
            pygame.K_3: 'boomerang',
            pygame.K_4: 'hookshot',
            pygame.K_5: 'lamp',
            pygame.K_6: 'hammer',
            pygame.K_7: 'fire_rod',
            pygame.K_8: 'ice_rod',
        }
        if key in item_keys:
            self.player_data['equipped_item'] = item_keys[key]

    def input(self):
        """Handle continuous input (movement)."""
        keys = pygame.key.get_pressed()

        # Movement
        self.direction.x = 0
        self.direction.y = 0

        if not self.attacking:
            # Vertical
            if keys[pygame.K_UP] or keys[pygame.K_w]:
                self.direction.y = -1
                self.facing = UP
            elif keys[pygame.K_DOWN] or keys[pygame.K_s]:
                self.direction.y = 1
                self.facing = DOWN

            # Horizontal
            if keys[pygame.K_LEFT] or keys[pygame.K_a]:
                self.direction.x = -1
                self.facing = LEFT
            elif keys[pygame.K_RIGHT] or keys[pygame.K_d]:
                self.direction.x = 1
                self.facing = RIGHT

        # Stop running if action key released
        if not (keys[pygame.K_l] or keys[pygame.K_c]):
            self.running = False

        # Spin attack charge (hold sword button)
        if keys[pygame.K_j] or keys[pygame.K_z]:
            if not self.attacking and not self.spin_attack_charging:
                self.spin_attack_charging = True
                self.spin_charge_start = pygame.time.get_ticks()
        else:
            if self.spin_attack_charging:
                # Check if fully charged
                if pygame.time.get_ticks() - self.spin_charge_start >= SPIN_ATTACK_CHARGE_TIME:
                    self.spin_attack()
                self.spin_attack_charging = False

    def attack(self):
        """Perform sword attack."""
        current_time = pygame.time.get_ticks()
        if current_time - self.attack_time >= self.attack_cooldown:
            self.attacking = True
            self.attack_time = current_time
            self.create_attack(self)

    def spin_attack(self):
        """Perform spin attack (charged sword)."""
        self.attacking = True
        self.attack_time = pygame.time.get_ticks()
        # Spin attack hits in all directions with more damage
        self.create_attack(self)

    def use_item(self):
        """Use currently equipped item."""
        if not self.can_use_item:
            return

        current_time = pygame.time.get_ticks()
        equipped = self.player_data['equipped_item']
        items = self.player_data['items']

        # Check magic cost
        magic_cost = ITEM_MAGIC_COST.get(equipped, 0)
        if self.magic < magic_cost:
            return

        if equipped == 'bombs' and self.player_data['bombs'] > 0:
            self.player_data['bombs'] -= 1
            self.create_projectile(self.rect.center, self.facing, 'bomb')
            self.can_use_item = False
            self.item_use_time = current_time

        elif equipped == 'bow' and items['bow'] and self.player_data['arrows'] > 0:
            self.player_data['arrows'] -= 1
            self.create_projectile(self.rect.center, self.facing, 'arrow')
            self.can_use_item = False
            self.item_use_time = current_time

        elif equipped == 'boomerang' and items['boomerang'] != 'none':
            self.create_projectile(self.rect.center, self.facing, 'boomerang')
            self.can_use_item = False
            self.item_use_time = current_time

        elif equipped == 'hookshot' and items['hookshot']:
            self.create_projectile(self.rect.center, self.facing, 'hookshot')
            self.can_use_item = False
            self.item_use_time = current_time

        elif equipped == 'fire_rod' and items['fire_rod']:
            self.magic -= magic_cost
            self.create_projectile(self.rect.center, self.facing, 'fire')
            self.can_use_item = False
            self.item_use_time = current_time

        elif equipped == 'ice_rod' and items['ice_rod']:
            self.magic -= magic_cost
            self.create_projectile(self.rect.center, self.facing, 'ice')
            self.can_use_item = False
            self.item_use_time = current_time

    def take_damage(self, amount, source_pos):
        """Take damage from an enemy or hazard."""
        if self.invincible:
            return

        self.health -= amount
        self.player_data['health'] = self.health

        # Knockback
        knockback_dir = pygame.math.Vector2(self.rect.center) - pygame.math.Vector2(source_pos)
        if knockback_dir.length() > 0:
            knockback_dir.normalize_ip()
        self.rect.x += knockback_dir.x * 20
        self.rect.y += knockback_dir.y * 20
        self.hitbox.center = self.rect.center

        # Start invincibility
        self.invincible = True
        self.invincible_time = pygame.time.get_ticks()

    def heal(self, amount):
        """Heal the player."""
        self.health = min(self.health + amount, self.max_health)
        self.player_data['health'] = self.health

    def restore_magic(self, amount):
        """Restore magic."""
        self.magic = min(self.magic + amount, self.max_magic)
        self.player_data['magic'] = self.magic

    def move(self, dt):
        """Move the player."""
        if self.direction.magnitude() != 0:
            self.direction = self.direction.normalize()

        speed = self.run_speed if self.running else self.speed

        # Horizontal movement
        self.hitbox.x += self.direction.x * speed * dt
        self.collision('horizontal')

        # Vertical movement
        self.hitbox.y += self.direction.y * speed * dt
        self.collision('vertical')

        # Sync rect with hitbox
        self.rect.center = self.hitbox.center

    def collision(self, direction):
        """Handle collision with obstacles."""
        for sprite in self.obstacle_sprites:
            if sprite.rect.colliderect(self.hitbox):
                if direction == 'horizontal':
                    if self.direction.x > 0:  # Moving right
                        self.hitbox.right = sprite.rect.left
                    elif self.direction.x < 0:  # Moving left
                        self.hitbox.left = sprite.rect.right

                elif direction == 'vertical':
                    if self.direction.y > 0:  # Moving down
                        self.hitbox.bottom = sprite.rect.top
                    elif self.direction.y < 0:  # Moving up
                        self.hitbox.top = sprite.rect.bottom

    def cooldowns(self):
        """Handle attack and item cooldowns."""
        current_time = pygame.time.get_ticks()

        # Attack cooldown
        if self.attacking:
            if current_time - self.attack_time >= self.attack_cooldown:
                self.attacking = False

        # Item cooldown
        if not self.can_use_item:
            if current_time - self.item_use_time >= self.item_cooldown:
                self.can_use_item = True

        # Invincibility
        if self.invincible:
            if current_time - self.invincible_time >= self.invincibility_duration:
                self.invincible = False

    def get_status(self):
        """Determine animation status."""
        direction_names = {UP: 'up', DOWN: 'down', LEFT: 'left', RIGHT: 'right'}
        dir_name = direction_names[self.facing]

        if self.attacking:
            self.status = f'attack_{dir_name}'
        elif self.direction.magnitude() != 0:
            self.status = f'walk_{dir_name}'
        else:
            self.status = f'idle_{dir_name}'

    def animate(self):
        """Animate the player sprite."""
        animation = self.animations.get(self.status, self.animations['idle_down'])

        current_time = pygame.time.get_ticks()
        if current_time - self.last_animation_time >= self.animation_speed:
            self.last_animation_time = current_time
            self.animation_index = (self.animation_index + 1) % len(animation)

        if animation:
            self.image = animation[int(self.animation_index) % len(animation)]

        # Flicker when invincible
        if self.invincible:
            if (current_time // 100) % 2:
                self.image.set_alpha(100)
            else:
                self.image.set_alpha(255)
        else:
            self.image.set_alpha(255)

    def update(self, dt):
        """Update player state."""
        self.input()
        self.cooldowns()
        self.get_status()
        self.move(dt)
        self.animate()

File: Phaser/Zelda/phaser/src/entities/Player.ts
// Player class - handles Link's movement, combat, and interactions

import Phaser from 'phaser';
import {
  PLAYER_SPEED,
  SWORD_COOLDOWN,
  SWORD_DAMAGE,
  SPIN_ATTACK_CHARGE_TIME,
  SPIN_ATTACK_DAMAGE,
  INVINCIBILITY_DURATION,
  Direction,
  DIRECTION_VECTORS,
  TILE_SIZE,
  SCALE,
  ITEM_DAMAGE,
  ITEM_MAGIC_COST,
} from '../config';
import { PlayerData } from '../types';

export class Player extends Phaser.Physics.Arcade.Sprite {
  public playerData: PlayerData;
  public health: number;
  public maxHealth: number;
  public magic: number;
  public maxMagic: number;
  public facing: Direction = Direction.DOWN;

  private cursors!: Phaser.Types.Input.Keyboard.CursorKeys;
  private wasdKeys!: { W: Phaser.Input.Keyboard.Key; A: Phaser.Input.Keyboard.Key; S: Phaser.Input.Keyboard.Key; D: Phaser.Input.Keyboard.Key };
  private swordKey!: Phaser.Input.Keyboard.Key;
  private swordKey2!: Phaser.Input.Keyboard.Key;
  private itemKey!: Phaser.Input.Keyboard.Key;
  private itemKey2!: Phaser.Input.Keyboard.Key;

  private isAttacking: boolean = false;
  private attackCooldown: number = 0;
  private spinCharging: boolean = false;
  private spinChargeStart: number = 0;

  public isInvincible: boolean = false;
  private invincibleTimer: number = 0;
  private flickerTimer: number = 0;

  private canUseItem: boolean = true;
  private itemCooldown: number = 0;

  // Shield blocking
  public isBlocking: boolean = false;
  private shieldKey!: Phaser.Input.Keyboard.Key;

  // Pegasus boots dash
  public isDashing: boolean = false;
  private dashSpeed: number = 450;
  private dashDuration: number = 400;
  private dashCooldown: number = 0;
  private dashTimer: number = 0;

  // Lamp
  private lampActive: boolean = false;
  public lampRadius: number = 150;

  // Lifted object
  public isCarrying: boolean = false;
  public carriedObject: Phaser.GameObjects.Sprite | null = null;

  // Callbacks for combat system
  public onAttack?: (player: Player, damage: number, isSpinAttack: boolean) => void;
  public onProjectile?: (x: number, y: number, direction: Direction, type: string) => void;
  public onLampUse?: (x: number, y: number, radius: number) => void;
  public onHammerUse?: (x: number, y: number, direction: Direction) => void;
  public onShieldBlock?: (blocked: boolean) => void;

  constructor(scene: Phaser.Scene, x: number, y: number, playerData: PlayerData) {
    super(scene, x, y, `player_idle_down`);

    this.playerData = playerData;
    this.health = playerData.health;
    this.maxHealth = playerData.maxHealth;
    this.magic = playerData.magic;
    this.maxMagic = playerData.maxMagic;
    this.facing = playerData.direction;

    // Add to scene
    scene.add.existing(this);
    scene.physics.add.existing(this);

    // Set up physics body
    this.setCollideWorldBounds(true);
    const body = this.body as Phaser.Physics.Arcade.Body;
    body.setSize(TILE_SIZE * 0.7, TILE_SIZE * 0.7);
    body.setOffset(TILE_SIZE * 0.15, TILE_SIZE * 0.3);

    // Scale sprite
    this.setScale(SCALE);

    // Set up input
    this.setupInput();
  }

  private setupInput(): void {
    const keyboard = this.scene.input.keyboard!;

    this.cursors = keyboard.createCursorKeys();
    this.wasdKeys = {
      W: keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.W),
      A: keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.A),
      S: keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.S),
      D: keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.D),
    };

    this.swordKey = keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.J);
    this.swordKey2 = keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.Z);
    this.itemKey = keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.K);
    this.itemKey2 = keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.X);
    this.shieldKey = keyboard.addKey(Phaser.Input.Keyboard.KeyCodes.SHIFT);

    // Quick item select (1-8)
    for (let i = 1; i <= 8; i++) {
      keyboard.addKey(48 + i).on('down', () => this.quickSelectItem(i));
    }
  }

  private quickSelectItem(slot: number): void {
    const items = ['bombs', 'bow', 'boomerang', 'hookshot', 'lamp', 'hammer', 'fire_rod', 'ice_rod'];
    if (slot <= items.length) {
      this.playerData.equippedItem = items[slot - 1];
    }
  }

  update(time: number, delta: number): void {
    if (this.isAttacking) {
      this.updateAttack(time, delta);
      return;
    }

    // Handle dashing
    if (this.isDashing) {
      this.updateDash(time, delta);
      return;
    }

    // Handle shield blocking
    this.handleShieldBlock();

    // Handle movement input (unless blocking)
    if (!this.isBlocking) {
      this.handleMovement();
    } else {
      // Stop when blocking
      (this.body as Phaser.Physics.Arcade.Body).setVelocity(0, 0);
    }

    // Handle combat input
    this.handleCombat(time);

    // Handle item usage
    this.handleItems(time);

    // Update cooldowns
    this.updateCooldowns(delta);

    // Update invincibility
    this.updateInvincibility(time, delta);

    // Update carried object position
    this.updateCarriedObject();

    // Update animation
    this.updateAnimation();

    // Sync player data
    this.syncPlayerData();
  }

  private handleShieldBlock(): void {
    const hasShield = this.playerData.items.shield !== 'none';
    const wasBlocking = this.isBlocking;

    this.isBlocking = hasShield && this.shieldKey.isDown && !this.isCarrying;

    // Callback on state change
    if (wasBlocking !== this.isBlocking && this.onShieldBlock) {
      this.onShieldBlock(this.isBlocking);
    }
  }

  private updateDash(time: number, delta: number): void {
    this.dashTimer -= delta;

    if (this.dashTimer <= 0) {
      this.isDashing = false;
      this.dashCooldown = 500; // Cooldown after dash
      return;
    }

    // Continue moving in dash direction
    const body = this.body as Phaser.Physics.Arcade.Body;
    const dirVec = DIRECTION_VECTORS[this.facing];
    body.setVelocity(dirVec.x * this.dashSpeed, dirVec.y * this.dashSpeed);
  }

  private updateCarriedObject(): void {
    if (this.isCarrying && this.carriedObject) {
      this.carriedObject.x = this.x;
      this.carriedObject.y = this.y - TILE_SIZE * SCALE;
      this.carriedObject.setDepth(this.depth + 1);
    }
  }

  public startDash(): void {
    if (!this.playerData.items.boots || this.dashCooldown > 0 || this.isDashing) {
      return;
    }

    this.isDashing = true;
    this.dashTimer = this.dashDuration;

    // Start dash velocity
    const body = this.body as Phaser.Physics.Arcade.Body;
    const dirVec = DIRECTION_VECTORS[this.facing];
    body.setVelocity(dirVec.x * this.dashSpeed, dirVec.y * this.dashSpeed);
  }

  public liftObject(obj: Phaser.GameObjects.Sprite): void {
    if (this.isCarrying) return;

    this.isCarrying = true;
    this.carriedObject = obj;
  }

  public throwCarriedObject(): void {
    if (!this.isCarrying || !this.carriedObject) return;

    // The interactable will handle the throw logic
    this.isCarrying = false;
    this.carriedObject = null;
  }

  private handleMovement(): void {
    const body = this.body as Phaser.Physics.Arcade.Body;
    let vx = 0;
    let vy = 0;

    // Vertical movement (priority over horizontal for LTTP feel)
    if (this.cursors.up.isDown || this.wasdKeys.W.isDown) {
      vy = -PLAYER_SPEED;
      this.facing = Direction.UP;
    } else if (this.cursors.down.isDown || this.wasdKeys.S.isDown) {
      vy = PLAYER_SPEED;
      this.facing = Direction.DOWN;
    }

    // Horizontal movement (only if not moving vertically, for 4-direction movement)
    if (vy === 0) {
      if (this.cursors.left.isDown || this.wasdKeys.A.isDown) {
        vx = -PLAYER_SPEED;
        this.facing = Direction.LEFT;
      } else if (this.cursors.right.isDown || this.wasdKeys.D.isDown) {
        vx = PLAYER_SPEED;
        this.facing = Direction.RIGHT;
      }
    }

    body.setVelocity(vx, vy);
  }

  private handleCombat(time: number): void {
    const swordPressed = this.swordKey.isDown || this.swordKey2.isDown;

    if (swordPressed && !this.isAttacking && this.attackCooldown <= 0) {
      if (!this.spinCharging) {
        // Start charging spin attack
        this.spinCharging = true;
        this.spinChargeStart = time;
      }
    } else if (!swordPressed && this.spinCharging) {
      // Released sword button
      const chargeTime = time - this.spinChargeStart;

      if (chargeTime >= SPIN_ATTACK_CHARGE_TIME) {
        // Spin attack!
        this.performSpinAttack();
      } else {
        // Normal attack
        this.performAttack();
      }

      this.spinCharging = false;
    }
  }

  private performAttack(): void {
    this.isAttacking = true;
    this.attackCooldown = SWORD_COOLDOWN;
    (this.body as Phaser.Physics.Arcade.Body).setVelocity(0, 0);

    // Play attack animation
    const dirName = this.getDirectionName();
    this.play(`player_attack_${dirName}`);

    // Notify combat system
    if (this.onAttack) {
      this.onAttack(this, SWORD_DAMAGE, false);
    }
  }

  private performSpinAttack(): void {
    this.isAttacking = true;
    this.attackCooldown = SWORD_COOLDOWN * 1.5;
    (this.body as Phaser.Physics.Arcade.Body).setVelocity(0, 0);

    // Play attack animation
    const dirName = this.getDirectionName();
    this.play(`player_attack_${dirName}`);

    // Notify combat system (spin attack hits all directions with more damage)
    if (this.onAttack) {
      this.onAttack(this, SPIN_ATTACK_DAMAGE, true);
    }
  }

  private updateAttack(time: number, delta: number): void {
    // Wait for attack animation to complete
    if (!this.anims.isPlaying) {
      this.isAttacking = false;
    }
  }

  private handleItems(time: number): void {
    if ((this.itemKey.isDown || this.itemKey2.isDown) && this.canUseItem) {
      this.useItem();
    }
  }

  private useItem(): void {
    const equipped = this.playerData.equippedItem;
    const items = this.playerData.items;

    // Check magic cost
    const magicCost = ITEM_MAGIC_COST[equipped] || 0;
    if (this.magic < magicCost) return;

    let used = false;

    switch (equipped) {
      case 'bombs':
        if (this.playerData.bombs > 0) {
          this.playerData.bombs--;
          this.fireProjectile('bomb');
          used = true;
        }
        break;
      case 'bow':
        if (items.bow && this.playerData.arrows > 0) {
          this.playerData.arrows--;
          this.fireProjectile('arrow');
          used = true;
        }
        break;
      case 'boomerang':
        if (items.boomerang !== 'none') {
          this.fireProjectile('boomerang');
          used = true;
        }
        break;
      case 'hookshot':
        if (items.hookshot) {
          this.fireProjectile('hookshot');
          used = true;
        }
        break;
      case 'fire_rod':
        if (items.fire_rod) {
          this.magic -= magicCost;
          this.fireProjectile('fire');
          used = true;
        }
        break;
      case 'ice_rod':
        if (items.ice_rod) {
          this.magic -= magicCost;
          this.fireProjectile('ice');
          used = true;
        }
        break;
      case 'lamp':
        if (items.lamp && this.magic >= 2) {
          this.magic -= 2;
          this.useLamp();
          used = true;
        }
        break;
      case 'hammer':
        if (items.hammer) {
          this.useHammer();
          used = true;
        }
        break;
      case 'boots':
        if (items.boots) {
          this.startDash();
          used = true;
        }
        break;
    }

    if (used) {
      this.canUseItem = false;
      this.itemCooldown = 150;  // Fast item usage
    }
  }

  private useLamp(): void {
    this.lampActive = true;

    // Create light effect at player position + direction offset
    const offset = DIRECTION_VECTORS[this.facing];
    const lampX = this.x + offset.x * TILE_SIZE * SCALE;
    const lampY = this.y + offset.y * TILE_SIZE * SCALE;

    if (this.onLampUse) {
      this.onLampUse(lampX, lampY, this.lampRadius);
    }

    // Also fire a small fire projectile
    this.fireProjectile('fire');

    // Lamp stays lit for a bit
    this.scene.time.delayedCall(3000, () => {
      this.lampActive = false;
    });
  }

  private useHammer(): void {
    // Stop movement during hammer use
    (this.body as Phaser.Physics.Arcade.Body).setVelocity(0, 0);

    // Calculate hammer hit position
    const offset = DIRECTION_VECTORS[this.facing];
    const hammerX = this.x + offset.x * TILE_SIZE * SCALE;
    const hammerY = this.y + offset.y * TILE_SIZE * SCALE;

    if (this.onHammerUse) {
      this.onHammerUse(hammerX, hammerY, this.facing);
    }

    // Also does sword damage in a small area
    if (this.onAttack) {
      this.onAttack(this, ITEM_DAMAGE['hammer'] || 4, false);
    }

    // Small screen shake for hammer impact
    this.scene.cameras.main.shake(100, 0.01);
  }

  private fireProjectile(type: string): void {
    if (this.onProjectile) {
      const offset = DIRECTION_VECTORS[this.facing];
      const projX = this.x + offset.x * TILE_SIZE * SCALE;
      const projY = this.y + offset.y * TILE_SIZE * SCALE;
      this.onProjectile(projX, projY, this.facing, type);
    }
  }

  private updateCooldowns(delta: number): void {
    if (this.attackCooldown > 0) {
      this.attackCooldown -= delta;
    }

    if (this.itemCooldown > 0) {
      this.itemCooldown -= delta;
      if (this.itemCooldown <= 0) {
        this.canUseItem = true;
      }
    }

    if (this.dashCooldown > 0) {
      this.dashCooldown -= delta;
    }
  }

  private updateInvincibility(time: number, delta: number): void {
    if (this.isInvincible) {
      this.invincibleTimer -= delta;
      this.flickerTimer += delta;

      // Flicker effect
      if (this.flickerTimer > 50) {
        this.setAlpha(this.alpha === 1 ? 0.3 : 1);
        this.flickerTimer = 0;
      }

      if (this.invincibleTimer <= 0) {
        this.isInvincible = false;
        this.setAlpha(1);
      }
    }
  }

  private updateAnimation(): void {
    if (this.isAttacking) return;

    const dirName = this.getDirectionName();
    const body = this.body as Phaser.Physics.Arcade.Body;

    if (body.velocity.x !== 0 || body.velocity.y !== 0) {
      // Walking
      const animKey = `player_walk_${dirName}`;
      if (this.anims.currentAnim?.key !== animKey) {
        this.play(animKey);
      }
    } else {
      // Idle
      this.setTexture(`player_idle_${dirName}`);
      this.anims.stop();
    }
  }

  private getDirectionName(): string {
    switch (this.facing) {
      case Direction.UP: return 'up';
      case Direction.DOWN: return 'down';
      case Direction.LEFT: return 'left';
      case Direction.RIGHT: return 'right';
    }
  }

  public takeDamage(amount: number, sourceX: number, sourceY: number): boolean {
    if (this.isInvincible) return false;

    // Check shield block
    if (this.isBlocking) {
      // Calculate if attack is from the front (blocked)
      const angleToSource = Phaser.Math.Angle.Between(this.x, this.y, sourceX, sourceY);
      const facingAngle = this.getFacingAngle();
      const angleDiff = Math.abs(Phaser.Math.Angle.Wrap(angleToSource - facingAngle));

      // Block if attack is within ~90 degrees of facing direction
      if (angleDiff < Math.PI / 2) {
        // Blocked! Small knockback but no damage
        const angle = Phaser.Math.Angle.Between(sourceX, sourceY, this.x, this.y);
        const knockbackForce = 100; // Reduced knockback
        const body = this.body as Phaser.Physics.Arcade.Body;
        body.setVelocity(
          Math.cos(angle) * knockbackForce,
          Math.sin(angle) * knockbackForce
        );
        return false;
      }
    }

    this.health -= amount;
    this.playerData.health = this.health;

    // Knockback
    const angle = Phaser.Math.Angle.Between(sourceX, sourceY, this.x, this.y);
    const knockbackForce = 200;
    const body = this.body as Phaser.Physics.Arcade.Body;
    body.setVelocity(
      Math.cos(angle) * knockbackForce,
      Math.sin(angle) * knockbackForce
    );

    // Start invincibility
    this.isInvincible = true;
    this.invincibleTimer = INVINCIBILITY_DURATION;
    this.flickerTimer = 0;

    // Check for death
    if (this.health <= 0) {
      this.die();
    }

    return true;
  }

  private getFacingAngle(): number {
    switch (this.facing) {
      case Direction.UP: return -Math.PI / 2;
      case Direction.DOWN: return Math.PI / 2;
      case Direction.LEFT: return Math.PI;
      case Direction.RIGHT: return 0;
    }
  }

  public heal(amount: number): void {
    this.health = Math.min(this.health + amount, this.maxHealth);
    this.playerData.health = this.health;
  }

  public restoreMagic(amount: number): void {
    this.magic = Math.min(this.magic + amount, this.maxMagic);
    this.playerData.magic = this.magic;
  }

  private die(): void {
    // Emit death event - GameScene will handle transition to GameOverScene
    this.scene.events.emit('playerDeath');
  }

  private syncPlayerData(): void {
    this.playerData.position.x = this.x;
    this.playerData.position.y = this.y;
    this.playerData.direction = this.facing;
    this.playerData.health = this.health;
    this.playerData.magic = this.magic;
  }

  public getAttackHitbox(): Phaser.Geom.Rectangle {
    const size = TILE_SIZE * SCALE;
    const offset = DIRECTION_VECTORS[this.facing];

    return new Phaser.Geom.Rectangle(
      this.x + offset.x * size - size / 2,
      this.y + offset.y * size - size / 2,
      size,
      size
    );
  }

  public getSpinAttackHitbox(): Phaser.Geom.Rectangle {
    const size = TILE_SIZE * SCALE * 2.5;
    return new Phaser.Geom.Rectangle(
      this.x - size / 2,
      this.y - size / 2,
      size,
      size
    );
  }
}
